packages/@webex/internal-plugin-metrics/src/call-diagnostic/call-diagnostic-metrics.ts

```diff
@@ -525,4 +549,42 @@ export default class CallDiagnosticMetrics extends StatelessWebexPlugin {
 
     return this.callDiagnosticEventsBatcher.request(finalEvent);
   }
+
+  /**
+   * Prepare Client Event CA event.
+   * @param arg - submit params
+   * @param arg.event - event key
+   * @param arg.payload - additional payload to be merged with default payload
+   * @param arg.options - payload
+   * @returns {Promise<any>}
+   * @throws
+   */
+  public async prepareMetricFetchOptions({
```

Comment:
@chburket  can you explain more on how this work , isn't there a way we could send this metic on the SDK side too? we would be happy to help it

cc @sreenara  @mkesavan13

Reply:
@arun3528, I used a similar strategy for CSAT surveys when the browser is closed.

There is a browser window event that we can listen for (` beforeUnload`) that fires when the browser tab is refreshed or closed, and when the browser is closed. Requests can be made from here but the `fetch` API must be used because it has a `keepalive` option that allows the request to outlive the tab.

In addition, we've found that there is only a small window for making these requests. Calling async functions, dispatching actions, etc. will often not complete when the browser itself is closed (otherwise we could call `webex.internal.newMetrics.submitClientEvent(args)` directly here).

So the idea is to pre-gather all the information/options needed for the request(s), and then simply and quickly fire the fetch(es) when `beforeUnload` is triggered.

We will do this by calling `webex.internal.newMetrics.prepareFetchOptions(args)` from a hook in Cantina when a meeting is active, and generate a function to "fetch" the metric data for immediate use when `beforeUnload` fires. In addition, this hook will add/remove the `beforeUnload` event handler from `window` as needed (e.g. add immediately after meeting join, and remove when user leaves).

I assumed the `window.on('beforeUnload')` logic was a better fit in Cantina, but please let me know if that is not the case. The SDK `prepareFetchOptions()` function could be made to return the function to call when browser is closing (e.g. `prepareFetchFunction`), but it's useful to be able to tell if the options changed so we know when to update the `beforeUnload` listener. We could definitely create an SDK function to do the actual `fetch` but that is one line of code.

There *will* be other code that will need to be run immediately before the fetch is called (updating the `originTime` values) so it would make some sense for that logic to be in the SDK, but it's pretty simple code and we can refactor later, as needed.

So anyway, that's the general idea. I'm happy to hear better ideas! As I said, this strategy is already in use for CSAT surveys, although more is done in Cantina because more of the old metrics code is in Cantina (vs the SDK).

Reply:
@torgeadelin and I are going to discuss offline

Reply:
Also important to note that the event won't have the correct triggerTime and needs appending at the time of fetch. Worth mentioning in the comment as well.

Reply:
Added comments